### PR TITLE
Fix tray icon not displaying "Open main dialog"

### LINKED ISSUE #4476: [Bug]: main (blue background) UI gone in 3.5 under Gnome/Wayland/Ubuntu 22.04
### ⚠️ Before submitting, please verify the following: ⚠️

- [X] This is a **bug**, not a question or a configuration issue.
- [X] This issue is **not** already reported on Github (I've searched it).
- [X] Nextcloud Server and Desktop Client are **up to date**. See [Server Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) and [Desktop Releases](https://nextcloud.com/install/#install-clients) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/)

### Bug description

Previously there used to be a funny little popup window thing, which I think is considered the main UI, as pictured below:

screenshot

Now there isn't.If I click the tray icon I get a menu (Add account, Pause sync for all, ... Settings...) and the only UI I can get to is the Settings screen. (Previously, in 3.4 it was called "Open main dialog")

The main UI did provide a few features that I can't live without; e.g. re conficts.

I've tried all types of clicking on the icon but can't get it to appear any more. If that UI was gone forever because the functionality had been moved into the (much better behaved) settings window that would be fine, but I don't think that's what's happened; I think it's a bug that it's missing?

### Steps to reproduce

Described above.

### Expected behavior

There should be some way to make the main UI show up. e.g. put a "Show main window" (or some other phrase) in the popup menu, alongside Settings etc. Or double-click or something.

### Which files are affected by this bug

N/A

### Operating system

Linux

### Which version of the operating system you are running.

Ubuntu 22.04

### Package

Appimage 3.5.0 "ubuntu-5.15.0-25-generic"

### Nextcloud Server version

23.0.4

### Is this bug present after an update or on a fresh install?

Updated to a major version (ex. 3.3.6 to 3.4.0) 3.4 → 3.5

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### Are you using an external user-backend?

- [ ] Default internal user-backend
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Nextcloud Server logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by claucambra**
This PR fixes a bug where the tray context menu never displays "Open main dialog", only "Add account", making the tray inoperable on GNOME.

Fixes #4476

### CODE REVIEW

**Review by camilasan — APPROVED** [review]

File: src/gui/systray.h [hunk 1]
```diff
@@ -94,7 +94,7 @@ class Systray
     void shutdown();
 
     void hideWindow();
-    void showWindow();
+    void showWindow(); // Listened to by the QML Window component
```
Comment (mgallien) [inline src/gui/systray.h:97]: this change seems unrelated

File: src/gui/systray.cpp [hunk 2]
```diff
@@ -160,6 +136,40 @@ void Systray::create()
     }
 }
 
+void Systray::setupContextMenu()
+{
+    _contextMenu = new QMenu();
```
Comment (mgallien) [inline src/gui/systray.cpp:141]: once `_contextMenu` has been already set to a valid pointer, executing this line will result in a memory leak if I understand the behavior of `QPointer` correctly.

what you want is either `std::auto_ptr` or `QScopedPointer`

**Review by mgallien — COMMENTED** [review]
looks good, see my comments

File: src/gui/systray.h (hunk 1 above)
Comment (claucambra) [reply src/gui/systray.h:97]: I wracked my head for a while trying to figure out how this signal invoked the QML window until realising there was a connection in `Window.qml`. Can remove, but might prove useful for future contributors

File: src/gui/systray.cpp (hunk 2 above)
Comment (claucambra) [reply src/gui/systray.cpp:142]: Both of these alternative wrappers cause an immediate segfault on application start

Comment (claucambra) [reply src/gui/systray.cpp:142]: I have added a quick `_contextMenu.clear()` before the assignment which should resolve the memory leak issue?

File: src/gui/systray.h (hunk 1 above)
Comment (mgallien) [reply src/gui/systray.h:97]: at least, makes it a separate commit
if we need to revert the fix of the tray icon that would also remove your comment
that is the main reason to not add unrelated changes
best would be a separate pr

Comment (claucambra) [reply src/gui/systray.h]: That is fair -- removed the comment

File: src/gui/systray.cpp (hunk 2 above)
Comment (claucambra) [reply src/gui/systray.cpp:146]: Original QMenu now gets deallocated without any explosions

File: src/gui/systray.cpp [hunk 3]
```diff
@@ -164,6 +140,46 @@ void Systray::create()
...
+        const auto anyPaused = std::any_of(std::cbegin(folders), std::cend(folders), [](Folder *f) { return f->syncPaused(); });
+        const auto resumeText = folders.size() > 1 ? tr("Resume sync for all") : tr("Resume sync");
+        resumeAction->setText(resumeText);
+        resumeAction->setVisible(anyPaused);
+        resumeAction->setEnabled(anyPaused);
+    });
+
+    // If we delete the old QMenu before setting the new one the client will crash on GNOME.
+    if(oldContextMenu) {
+        delete oldContextMenu;
```
Comment (allexzander) [inline src/gui/systray.cpp:179]: @claucambra Can you do it in the place where you assign to `oldContextMenu` by using `QObject::deleteLater()` https://doc.qt.io/qt-5/qobject.html#deleteLater? Should do the same thing but, without this extra code at the end.

**Review by allexzander — APPROVED** [review]
@claucambra Looks good. See my comment.

File: src/gui/systray.cpp (hunk 3 above)
Comment (claucambra) [reply src/gui/systray.cpp]: Works great! TIL :)

**Pull request merged**